lib/pdf_attachment_reporter.rb

```diff
@@ -0,0 +1,179 @@
+      ]
+
+      live_organisation_names.each do |organisation_name|
+        csv << [
+          organisation_name,
+          live_organisation_published_pdfs_total_counts_hash[organisation_name],
+          live_organisation_published_pdfs_since_first_period_counts_hash[organisation_name],
+          live_organisation_published_pdfs_since_second_period_counts_hash[organisation_name]
+        ]
+      end
+    end
+  end
+
+private
+
+  class PDFAttachmentData
+    attr_reader :owning_organisation_name, :created_at
+
+    def initialize(owning_organisation_name, created_at)
+      @owning_organisation_name = owning_organisation_name
+      @created_at = created_at
+    end
+  end
+
+  def find_pdf_attachment_data(attachment)
+    attachable_object = find_attachable_object(attachment)
+
+    if attachable_object
+      if attachable_object.class == PolicyGroup
+        PDFAttachmentData.new(POLICY_GROUPS, attachment.created_at)
+      else
+        # Responses are only sometimes linked to organisations (via a consultation)
+        if attachable_object.is_a?(Response)
+          if attachable_object.consultation
+            last_published_edition_with_attachment = find_last_published_edition_with_attachment(attachable_object.consultation, attachment.attachment_data)
+
+            if last_published_edition_with_attachment
+              pdf_owning_organisation = guess_organisation_owner_of_edition(last_published_edition_with_attachment)
+              first_published_edition_with_attachment = find_first_published_version_with_attachment(attachable_object.consultation, attachment.attachment_data)
+            end
+          end
+        else
+          last_published_edition_with_attachment = find_last_published_edition_with_attachment(attachable_object, attachment.attachment_data)
+
+          if last_published_edition_with_attachment
+            pdf_owning_organisation = guess_organisation_owner_of_edition(last_published_edition_with_attachment)
+            first_published_edition_with_attachment = find_first_published_version_with_attachment(attachable_object, attachment.attachment_data)
+          end
+        end
+
+        if pdf_owning_organisation && pdf_owning_organisation.live? &&
+            first_published_edition_with_attachment && first_published_edition_with_attachment.public_timestamp
+          PDFAttachmentData.new(pdf_owning_organisation.name, first_published_edition_with_attachment.public_timestamp)
+        end
+      end
+    end
+  end
+
+  def find_attachable_object(attachment)
+    attachable_type = attachment.attachable_type.constantize
+    attachable_type.find_by_id(attachment.attachable_id)
+  end
+
+  def find_last_published_edition_with_attachment(edition, attachment_data)
+    edition.document.editions.order('created_at DESC').detect do |ed|
+      Edition::POST_PUBLICATION_STATES.include?(ed.state) && ed.attachments.any? { |a| a.attachment_data_id == attachment_data.id }
```

Comment:
👍  I should have searched for the constant before I went ahead with duplicating this work. See https://github.com/alphagov/whitehall/blob/dadc7f30af1e68a422e23ec046b6c8f5170b3226/lib/pdf_attachment_reporter.rb#L121